src/dialogs/editor.tsx

```diff
@@ -0,0 +1,203 @@
+    }
+
+    modify(content: string) {
+        this.update({ content, modified: true });
+    }
+
+    load_file() {
+        // Can't do this async because we can't get the tag via await
+        this.file.read()
+                .then(((content: string, tag: string) => {
+                    this.update({ content, tag_now: tag, tag_at_load: tag });
+                }) as any /* eslint-disable-line @typescript-eslint/no-explicit-any */)
+                .catch(error => {
+                    this.update({ error: cockpit.message(error) });
+                });
+    }
+
+    constructor(filename: string) {
+        super();
+        this.file = cockpit.file(filename);
+        this.state = new EditorState();
+
+        this.load_file();
+
+        cockpit.spawn(['test', '-w', filename])
+                .then(() => this.update({ writable: true }))
+                .catch(() => this.update({ writable: false }));
+
+        this.file.watch((_content, tag_now) => {
+            this.update({ tag_now });
+        }, { read: false });
+    }
+
+    async save() {
+        if (!this.state.tag_now)
+            return;
+
+        try {
+            this.update({ saving: true });
+            const tag = await this.file.replace(this.state.content, this.state.tag_now);
+            this.update({ saving: false, modified: false, tag_now: tag, tag_at_load: tag });
+        } catch (exc: any /* eslint-disable-line @typescript-eslint/no-explicit-any */) {
+            this.update({ error: cockpit.message(exc), saving: false });
+        }
+    }
+
+    close() {
+        this.file.close();
+    }
+}
+
+export const EditFileModal = ({ dialogResult, path } : {
+    dialogResult: DialogResult<void>,
+    path: string
+}) => {
+    const [state, setState] = React.useState(new EditorState());
+    const [editor, setEditor] = React.useState<Editor | null>(null);
+
+    React.useEffect(() => {
+        const editor = new Editor(path);
+        editor.on('updated', setState);
+        setEditor(editor);
+        return () => {
+            editor.close();
+        };
+    }, [path]);
+
+    const modified = state.modified;
+    React.useEffect(() => {
+        const before_unload = (event: BeforeUnloadEvent) => {
+            event.preventDefault();
+        };
+        if (modified) {
+            window.addEventListener('beforeunload', before_unload);
+            return () => {
+                window.removeEventListener('beforeunload', before_unload);
+            };
+        }
+    }, [modified]);
+
+    /* Translators: $0 represents a filename */
+    const title = cockpit.format(state?.writable ? _("Editing “$0”") : _("Viewing “$0”"), path);
+
+    return (
+        <Modal
+          position="top"
+          title={(modified ? "⭐ " : "") + title}
+          isOpen
+          onClose={() => dialogResult.resolve()}
+          variant={ModalVariant.large}
+          className='file-editor-modal'
+          footer={
+              state?.writable &&
```

Comment:
The dialog should have a "Close" button for non-writable files.

Reply:
Fixed.